src/ports/postgres/modules/graph/wcc.py_in

```diff
@@ -0,0 +1,329 @@
+        # looking at its neighbors are updated in the next table. At every
+        # iteration update only those nodes whose component_id in the previous
+        # iteration are greater than what was found in the current iteration.
+        plpy.execute("DROP TABLE IF EXISTS {0}".format(oldupdate))
+        plpy.execute("""
+            CREATE TEMP TABLE {oldupdate} AS
+            SELECT {message}.{vertex_id},
+                    MIN({message}.{component_id}) AS {component_id}
+            FROM {message}
+            GROUP BY {vertex_id}
+            {distribution}
+        """.format(**locals()))
+
+        plpy.execute("DROP TABLE IF EXISTS {0}".format(toupdate))
+        plpy.execute("""
+            CREATE TEMP TABLE {toupdate} AS
+            SELECT {oldupdate}.{vertex_id}, {oldupdate}.{component_id}
+            FROM {oldupdate}, {newupdate}
+            WHERE {oldupdate}.{vertex_id}={newupdate}.{vertex_id}
+                AND {oldupdate}.{component_id}<{newupdate}.{component_id}
+            {distribution}
+        """.format(**locals()))
+
+        if is_hawq:
+            plpy.execute("""
+                    TRUNCATE TABLE {temp_out_table};
+                    INSERT INTO {temp_out_table}
+                        SELECT *
+                        FROM {newupdate}
+                        WHERE NOT EXISTS (
+                            SELECT *
+                            FROM {toupdate}
+                            WHERE {newupdate}.{vertex_id}={toupdate}.{vertex_id}
+                        )
+                        UNION
+                        SELECT * FROM {toupdate};
+                """.format(**locals()))
+            plpy.execute("DROP TABLE {0}".format(newupdate))
+            plpy.execute("ALTER TABLE {0} RENAME TO {1}".format(temp_out_table,
+                            newupdate))
+            plpy.execute("""
+                    CREATE TABLE {temp_out_table} AS
+                    SELECT * FROM {newupdate}
+                    LIMIT 0
+                    {distribution}
+                """.format(**locals()))
+        else:
+            plpy.execute("""
+                    UPDATE {newupdate} SET
+                    {component_id}={toupdate}.{component_id}
+                    FROM {toupdate}
+                    WHERE {newupdate}.{vertex_id}={toupdate}.{vertex_id}
+                """.format(**locals()))
+
+        plpy.execute("DROP TABLE IF EXISTS {0}".format(message))
+        plpy.execute("""
+            CREATE TEMP TABLE {message} AS
+            SELECT {vertex_id}, MIN({component_id}) AS {component_id}
+            FROM (
+                SELECT {edge_table}.{src} AS {vertex_id}, {toupdate}.{component_id}
+                FROM {toupdate}, {edge_table}
+                WHERE {edge_table}.{dest} = {toupdate}.{vertex_id}
+                UNION ALL
+                SELECT {edge_table}.{dest} AS {vertex_id}, {toupdate}.{component_id}
+                FROM {toupdate}, {edge_table}
```

Comment:
What this query is trying to do is the following:
For each node, find all its neighbors and assign this node's `component_id` as their `component_id`. The `MIN()` in the outer `SELECT` clause will then choose the minimum of all possible neighbors' component_id and assign it to a node.
I don't think the suggestion in this comment will help me do it any more efficiently. We will still have to do the joins, since the suggested query will only find all edges in the `edge_table` that have `{vertex_id}` in either src or dest. Hence assuming a no-op on this.